controllers/cosmosdbsqldatabase_controller_test.go

```diff
@@ -0,0 +1,194 @@
+	assert.Equal(nil, err, "getting cosmos SQL DB from k8s")
+
+	// Update the Cosmos SQL DB throughput
+	throughPutRUs = 1000
+	sqlDBInstance.Spec.Throughput = &throughPutRUs
+	err = tc.k8sClient.Update(ctx, sqlDBInstance)
+	assert.Equal(nil, err, "updating cosmos sql DB in k8s")
+
+	assert.Eventually(func() bool {
+		throughputMatches, err := doesThroughputMatch(
+			ctx,
+			cosmosDBSQLDatabaseClient,
+			tc.resourceGroupName,
+			cosmosDBAccountName,
+			name,
+			throughPutRUs)
+		assert.Equal(nil, err, "err ensuring throughput matches")
+
+		return throughputMatches
+	}, tc.timeout, tc.retry, "wait for cosmos SQL DB throughput to be updated in Azure")
+
+	EnsureDelete(ctx, t, tc, sqlDBInstance)
+}
+
+func CosmosDBSQLDatabase_FailedThroughputUpdate(t *testing.T, cosmosDBAccountName string) {
+	t.Parallel()
+	defer PanicRecover(t)
+	ctx := context.Background()
+	assert := require.New(t)
+
+	name := GenerateTestResourceNameWithRandom("cosmosdbsql", 8)
+	namespace := "default"
+
+	sqlDBInstance := &v1alpha1.CosmosDBSQLDatabase{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: v1alpha1.CosmosDBSQLDatabaseSpec{
+			ResourceGroup:   tc.resourceGroupName,
+			CosmosDBAccount: cosmosDBAccountName,
+		},
+	}
+
+	// Create the Cosmos SQL DB
+	EnsureInstance(ctx, t, tc, sqlDBInstance)
+
+	// Get the updated Cosmos SQL DB from k8s
+	namespacedName := types.NamespacedName{Name: name, Namespace: namespace}
+	err := tc.k8sClient.Get(ctx, namespacedName, sqlDBInstance)
+	assert.Equal(nil, err, "getting cosmos SQL DB from k8s")
+
+	// Update the Cosmos SQL DB throughput (we expect that this will be rejected by Cosmos DB)
+	var throughPutRUs int32 = 1000
+	sqlDBInstance.Spec.Throughput = &throughPutRUs
+	err = tc.k8sClient.Update(ctx, sqlDBInstance)
+	assert.Equal(nil, err, "updating cosmos sql DB in k8s")
+
+	// Ensure we get an error eventually
+	assert.Eventually(func() bool {
+		updatedInstance := &v1alpha1.CosmosDBSQLDatabase{}
+		err = tc.k8sClient.Get(ctx, namespacedName, updatedInstance)
+		assert.Equal(nil, err, "err getting Cosmos SQL DB from k8s")
+
+		return updatedInstance.Status.Provisioned == false &&
+			updatedInstance.Status.FailedProvisioning == true &&
+			strings.Contains(updatedInstance.Status.Message, "Throughput update is not supported for resources created without dedicated throughput")
+	}, tc.timeout, tc.retry, "wait for sql database to be updated in k8s")
+
+	EnsureDelete(ctx, t, tc, sqlDBInstance)
+}
+
+func doesThroughputMatch(
+	ctx context.Context,
+	client documentdb.SQLResourcesClient,
+	rgName string,
+	cosmosDBAccountName string,
+	dbName string,
+	expectedThroughput int32) (bool, error) {
+
+	throughputSettings, err := client.GetSQLDatabaseThroughput(ctx, rgName, cosmosDBAccountName, dbName)
+	if err != nil {
+		return false, errors.Wrap(err, "err getting cosmos sql DB throughput from Azure")
+	}
+
+	if throughputSettings.ThroughputSettingsGetProperties == nil {
```

Comment:
The way it works in C# basically means that any nil short-circuits the rest of the checks to the right and just returns the nil. I'm not sure exactly how it'd work with `*` though. Even if it didn't, this could become:
```
actual := throughputSettings.ThroughputSettingsGetProperties?.Resource?.Throughput?

if actual != nil {
    return *actual == expected
}

return false
```

In the actual diffing code, a lot of that coalescing could be turned into:
`a?.b?.c == d?.e?.f` too